test/http_parser_test.dart

```diff
@@ -0,0 +1,811 @@
+// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library dart.http;
+
+import "dart:async";
+import "dart:isolate";
+import "dart:math";
+import "dart:typed_data";
+
+import 'package:http_io/src/http_headers_impl.dart';
+import 'package:http_io/src/http_parser.dart';
+import 'package:test/test.dart';
+
+class HttpParserTest {
+  static void runAllTests() {
+    test('parseRequest', testParseRequest);
+    test('parseResponse', testParseResponse);
+    test('parseInvalidRequest', testParseInvalidRequest);
+    test('parseInvalidResponse', testParseInvalidResponse);
+  }
+
+  static void _testParseRequest(
+      String request, String expectedMethod, String expectedUri,
+      {int expectedTransferLength: 0,
+      int expectedBytesReceived: 0,
+      Map<String, String> expectedHeaders: null,
+      bool chunked: false,
+      bool upgrade: false,
+      int unparsedLength: 0,
+      bool connectionClose: false,
+      String expectedVersion: "1.1"}) {
+    StreamController<List<int>> controller;
+    void reset() {
+      HttpParser httpParser = new HttpParser.requestParser();
+      controller = new StreamController(sync: true);
+      var port1 = new ReceivePort();
+      var port2 = new ReceivePort();
+
+      String method;
+      Uri uri;
+      HttpHeadersImpl headers;
+      int bytesReceived;
+      int unparsedBytesReceived;
+      bool upgraded;
+
+      httpParser.listenToStream(controller.stream);
+      httpParser.listen((incoming) {
+        method = incoming.method;
+        uri = incoming.uri;
+        headers = incoming.headers;
+        upgraded = incoming.upgraded;
+        expect(upgrade, equals(upgraded));
+
+        if (!chunked) {
+          expect(expectedTransferLength, equals(incoming.transferLength));
+        } else {
+          expect(-1, equals(incoming.transferLength));
+        }
+        if (expectedHeaders != null) {
+          expectedHeaders.forEach((String name, String value) =>
+              expect(value, equals(headers[name][0])));
+        }
+        incoming.listen((List<int> data) {
+          expect(upgraded, isFalse);
+          bytesReceived += data.length;
+        }, onDone: () {
+          port2.close();
+          expect(expectedMethod, equals(method));
+          expect(expectedUri, equals(uri.toString()));
+          expect(expectedVersion, equals(headers.protocolVersion));
```

Comment:
Is there any particular reason that `protocolVersion` isn't exposed through `HttpHeaders`?

Reply:
I don't know. If it looks like it should be there, then we can add it.

Reply:
It's a public member of `HttpHeadersImpl` and could be useful information. I can expose it in a separate PR.

Reply:
sgtm

Reply:
@zanderso , done.